src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1391 @@
+        ///
+        /// # Identifying sets of cases to run reachability analysis on:
+        /// For patterns that contain a disjunction `... or ...` we're going to perform reachability analysis for each branch of the `or`.
+        /// We effectively pick each analyzable `or` sequence in turn and expand it to top-level cases.
+        ///
+        /// For example `A and (B or C)` we'll check the reachability of two cases: `case A and B` and `case A and C`.
+        ///
+        /// Similarly, for `(A or B) and (C or D)` we'll check the reachability of two sets of two cases:
+        ///   1. { `case A`, `case B` } (we can truncate later test since we only care about the reachability of `A` and `B` here)
+        ///   2. { `case (A or B) and C`, `case (A or B) and D` }
+        ///
+        /// Similarly, for `A or ((B or C) and D)` we'll check the reachability of two sets of cases:
+        ///   1. { `case A:`, `case ((B or C) and D):` }
+        ///   2. { `case A:`, `case B:`, `case C:` }
+        ///
+        /// Similarly, for `A or (B and (C or D))` we'll check the reachability of two sets of cases:
+        ///   1. { `case A:`, `case (B and (C or D)):`
+        ///   2. { `case A:`, `case B and C:`, `case B and D:` }
+        /// </summary>
+        internal static void CheckRedundantPatternsForIsPattern(
+            CSharpCompilation compilation,
+            SyntaxNode syntax,
+            BoundExpression inputExpression,
+            BoundPattern pattern,
+            BindingDiagnosticBag diagnostics)
+        {
+            if (pattern.HasErrors)
+            {
+                return;
+            }
+
+            LabelSymbol defaultLabel = new GeneratedLabelSymbol("isPatternFailure");
+            var builder = new DecisionDagBuilder(compilation, defaultLabel: defaultLabel, forLowering: false, BindingDiagnosticBag.Discarded);
+            BoundDagTemp rootIdentifier = BoundDagTemp.ForOriginalInput(inputExpression);
+            var redundantNodes = PooledHashSet<SyntaxNode>.GetInstance();
+
+            var noPreviousCases = ArrayBuilder<StateForCase>.GetInstance(0);
+            CheckOrAndAndReachability(noPreviousCases, patternIndex: 0, pattern: pattern, builder: builder, rootIdentifier: rootIdentifier, syntax: syntax, diagnostics: diagnostics, redundantNodes);
+            ReportRedundant(redundantNodes, diagnostics);
+
+            redundantNodes.Free();
+            noPreviousCases.Free();
+        }
+
+        /// <summary>
+        /// <see cref="CheckRedundantPatternsForIsPattern"/>
+        /// </summary>
+        internal static void CheckRedundantPatternsForSwitchExpression(
+            CSharpCompilation compilation,
+            SyntaxNode syntax,
+            BoundExpression inputExpression,
+            ImmutableArray<BoundSwitchExpressionArm> switchArms,
+            BindingDiagnosticBag diagnostics)
+        {
+            LabelSymbol defaultLabel = new GeneratedLabelSymbol("isPatternFailure");
```

Comment:
nit: perhaps label name should be adjusted to reflect that this is a switch expression.